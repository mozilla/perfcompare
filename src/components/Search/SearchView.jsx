import React from 'react';
import Box from '@mui/material/Box';
import Container from '@mui/material/Container';
import Typography from '@mui/material/Typography';

import SearchInput from './SearchInput';

export default function SearchView() {
  return (
    <Container maxWidth="lg">
      <Box>
        <Typography variant="h1" component="div" align="center" gutterBottom>
          PerfCompare
        </Typography>
      </Box>
      <SearchInput />
    </Container>
  );
}
