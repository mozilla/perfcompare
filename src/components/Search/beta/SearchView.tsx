import PerfCompareHeader from '../../Shared/beta/PerfCompareHeader';

function SearchViewBeta() {
  return (
    <section className='perfcompare-body'>
      <PerfCompareHeader />
    </section>
  );
}

export default SearchViewBeta;
