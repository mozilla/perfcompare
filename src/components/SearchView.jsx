import * as React from 'react';

import Box from '@mui/material/Box';
import CssBaseline from '@mui/material/CssBaseline';
import Container from '@mui/material/Container';
import Typography from '@mui/material/Typography';

import SearchInput from './SearchInput';

export default function SearchView() {
  return (
    <React.Fragment>
      <CssBaseline />
      <Container maxWidth="lg">
        <Box>
          <Typography variant="h1" component="div" align="center" gutterBottom>
            PerfCompare
          </Typography>
        </Box>
        <SearchInput />
      </Container>
    </React.Fragment>
  );
}
