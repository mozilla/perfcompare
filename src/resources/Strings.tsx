export const Strings = {
  metaData: {
    pageTitle: {
      search: 'PerfCompare - Search',
      results: 'PerfCompare - Results',
    },
  },
  components: {
    topBanner: {
      text: 'This is an unstable pre-release version. Some features may not yet be supported.',
      linkText: 'File a bug on Bugzilla.',
      href: 'https://bugzilla.mozilla.org/enter_bug.cgi?product=Testing&component=PerfCompare&status_whiteboard=[pcf]',
    },
    header: {
      title: 'PerfCompare',
      tagline:
        'Analyze results of performance tests to detect regressions and identify opportunities for improvement.',
      bgLink:
        'url(https://user-images.githubusercontent.com/18633586/223226187-83621473-af10-4f7b-a7b9-564378b55c9e.png)',
      darkMode: 'Dark mode',
      lightMode: 'Light mode',
      learnMore: 'Learn more',
    },
    searchDefault: {
      title: 'Compare with a base or over time',
      sharedCollasped: {
        revisions: {
          selectLabel: 'Revisions',
          placeholder: 'Search base by ID number or author email',
        },
        framkework: {
          selectLabel: 'Framework',
          tooltip:
            'The framework or test harness containing the test you want to examine.',
        },
        button: 'Compare',
      },
      base: {
        title: 'Compare with a base',
        tagline: 'Analyze differences between specific revisions.',
        img: 'https://user-images.githubusercontent.com/88336547/233237125-1534220b-c343-421a-9133-ce8f151cb979.png',
        imgDark:
          'https://user-images.githubusercontent.com/88336547/233250674-004d071a-7c23-40f4-b348-0687a3fef6e3.png',

        save: 'Save',
        cancel: 'Cancel',
        editIcon:
          'https://user-images.githubusercontent.com/88336547/259900252-462ac221-f4ff-4b1c-bb6b-5df906e8007d.svg',
        collapsed: {
          warnings: {
            comparison:
              'Comparing “try” repository to any repository aside from "try" is not recommended.',
          },
          base: {
            selectLabel: 'Base',
            inputPlaceholder: 'Search base by ID number or author email',
            tooltip:
              'The baseline revision (no changes) your revision will be compared against.',
          },
          revision: {
            selectLabel: 'Revisions',
            inputPlaceholder: 'Search revision by ID number or author email',
            tooltip:
              'Revisions (typically including your changes) to compare against the selected base revision.',
          },
        },
      },
      overTime: {
        title: 'Compare over time',
        tagline:
          'Analyze differences between revisions within a specific time range.',
        img: 'https://user-images.githubusercontent.com/88336547/233250659-4012551b-e07a-44ce-accb-242e29d31914.png',
        imgDark:
          'https://user-images.githubusercontent.com/88336547/233250642-7fd7c217-e72b-4375-9078-7ed2f99cb0f7.png',
        collapedSelectLabel: 'Time range',
      },
    },
    expandableRow: {
      singleRun: 'Only one run (consider more runs for greater confidence).',
      Low: "A value of 'low' suggests less confidence that there is a sustained, significant change between the two revisions.",
      Medium:
        "A value of 'med' indicates uncertainty that there is a significant change. If you haven't already, consider retriggering the job to be more sure.",
      High: "A value of 'high' indicates more confidence that there is a significant change, however you should check the historical record for the test by looking at the graph to be more sure (some noisy tests can provide inconsistent results).",
    },
    noResultsFound: {
      mainMessage: 'No results found',
      note: 'For the selected revision(s), no results when compared to the base revision.',
    },
    comparisonRevisionDropdown: {
      allRevisions: 'All revisions',
    },
  },
  errors: {
    warningText: 'Search must be a 12- or 40-character hash, or email address',
  },
};
