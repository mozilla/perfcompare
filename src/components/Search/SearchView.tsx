import ArrowForward from '@mui/icons-material/ArrowForward';
import Button from '@mui/material/Button';
import Container from '@mui/material/Container';
import Grid from '@mui/material/Grid';
import { useSnackbar, VariantType } from 'notistack';
import { connect } from 'react-redux';
import { useNavigate } from 'react-router-dom';

import { repoMap, featureNotSupportedError } from '../../common/constants';
import type { RootState } from '../../common/store';
import useFilterCompareResults from '../../hooks/useFilterCompareResults';
import { Strings } from '../../resources/Strings';
import { Revision } from '../../types/state';
import RevisionSearch from '../Shared/RevisionSearch';
import SelectedRevisionsTable from '../Shared/SelectedRevisionsTable';
import SearchViewInit from './SearchViewInit';

const strings = Strings.components.compareDefault.sharedCollasped;

function SearchView(props: SearchViewProps) {
  const navigate = useNavigate();
  const { clearFilters } = useFilterCompareResults();
  const { enqueueSnackbar } = useSnackbar();
  const warningVariant: VariantType = 'warning';

  const goToCompareResultsPage = (selectedRevisions: Revision[]) => {
    // TODO: remove this check once comparing without a base
    //  and comparing multiple revisions against a base is enabled
    if (selectedRevisions.length === 1 || selectedRevisions.length > 2) {
      enqueueSnackbar(featureNotSupportedError as string, {
        variant: warningVariant,
      });
      return;
    }
    const revs = selectedRevisions.map((rev) => rev.revision);
    const repos = selectedRevisions.map((rev) => repoMap[rev.repository_id]);
    clearFilters();
    navigate({
      pathname: '/compare-results',
      search: `?revs=${revs.join(',')}&repos=${repos.join(',')}`,
    });
  };

  const { selectedRevisions } = props;

  return (
    <Container className='perfcompare-body' maxWidth='lg'>
      {/* Component to fetch recent revisions on mount */}
      <SearchViewInit />
      <Grid item xs={12}>
        {selectedRevisions.length > 0 && (
          <SelectedRevisionsTable view='search' />
        )}
      </Grid>
      <Grid item className='compare-button-section'>
        {selectedRevisions.length > 0 && (
          <Button
            className='compare-button'
            variant='contained'
            onClick={() => goToCompareResultsPage(selectedRevisions)}
          >
            {strings.button}
            <ArrowForward className='compare-icon' />
          </Button>
        )}
      </Grid>
      <RevisionSearch view='search' />
    </Container>
  );
}

interface SearchViewProps {
  searchResults: Revision[];
  selectedRevisions: Revision[];
}

function mapStateToProps(state: RootState) {
  return {
    searchResults: state.search.searchResults,
    selectedRevisions: state.selectedRevisions.revisions,
  };
}

export default connect(mapStateToProps)(SearchView);
