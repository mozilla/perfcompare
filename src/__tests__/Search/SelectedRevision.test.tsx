import userEvent from '@testing-library/user-event';

import SearchView from '../../components/Search/SearchView';
import { Strings } from '../../resources/Strings';
import getTestData from '../utils/fixtures';
import {
  screen,
  within,
  renderWithRouter,
  FetchMockSandbox,
} from '../utils/test-utils';

function renderComponent() {
  renderWithRouter(<SearchView title={Strings.metaData.pageTitle.search} />);
}

describe('SelectedRevision', () => {
  const { testData } = getTestData();

  beforeEach(() => {
    (global.fetch as FetchMockSandbox).get(
      'glob:https://treeherder.mozilla.org/api/project/*/push/*',
      {
        results: testData,
      },
    );
  });

  it('should show the selected checked revisions once a result checkbox is clicked, and remove it when X button is clicked', async () => {
    // set delay to null to prevent test time-out due to useFakeTimers
    const user = userEvent.setup({ delay: null });

    renderComponent();
    // focus input to show results
    const searchInput = screen.getAllByRole('textbox')[0];
    await user.click(searchInput);

    const noArmsLeft = await screen.findByRole('button', {
      name: /you've got no arms left!/,
    });
    const noArmsLeftCheckbox = within(noArmsLeft).getByRole('checkbox');

    await user.click(noArmsLeft);
    expect(noArmsLeft).toHaveClass('item-selected');
    expect(noArmsLeftCheckbox).toBeChecked();
    expect(noArmsLeft.querySelector('.Mui-checked')).toBeInTheDocument();

    const selectedRevsContainer = screen.getByTestId('selected-rev-item');
    expect(selectedRevsContainer).toMatchSnapshot();

    const removeButton = document.querySelectorAll('[title="remove revision"]');

    const removeIcon = screen.getByTestId('close-icon');
    expect(removeIcon).toBeInTheDocument();
    expect(screen.getAllByTestId('selected-rev-item')[0]).toBeInTheDocument();

    await user.click(removeButton[0]);

    expect(screen.queryByTestId('selected-rev-item')).not.toBeInTheDocument();
  });

  it('should show warning icon on selected try revision when try base is compared with a non try repository', async () => {
    const user = userEvent.setup({ delay: null });
    renderComponent();

    const baseDropdown = screen.getByRole('button', { name: 'Base' });
    expect(baseDropdown).toHaveTextContent('try');

    const firstSearchInput = screen.getAllByPlaceholderText(
      'Search base by ID number or author email',
    )[0];
    await user.click(firstSearchInput);
    await user.click(
      await screen.findByRole('button', {
        name: /you've got no arms left!/,
      }),
    );

    const newDropdown = screen.getAllByRole('button', { name: 'Revisions' })[0];
    await user.click(newDropdown);
    const mozRepoItem = await screen.findByRole('option', {
      name: 'mozilla-central',
    });
    await user.click(mozRepoItem);
    const alertIcon = await screen.findByRole('img', {
      name: 'Comparing “try” repository to any repository aside from “try” is not recommended.',
    });
    expect(alertIcon).toBeInTheDocument();
  });
});
