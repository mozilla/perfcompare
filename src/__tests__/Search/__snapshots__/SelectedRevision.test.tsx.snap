// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SelectedRevision should show the selected checked revisions once a result checkbox is clicked, and remove it when X button is clicked 1`] = `
<li
  class="MuiListItem-root MuiListItem-gutters MuiListItem-padding item-container item-0 item-base css-vtiz8x-MuiListItem-root"
  data-testid="selected-rev-item"
>
  <input
    name="baseRev"
    type="hidden"
    value="coconut"
  />
  <input
    name="baseRepo"
    type="hidden"
    value="try"
  />
  <div
    class="repo_f1bx8by5"
  >
    <div>
      try
    </div>
  </div>
  <div
    class="MuiButtonBase-root MuiListItemButton-root MuiListItemButton-gutters MuiListItemButton-root MuiListItemButton-gutters listItemButton_f5s7lzh css-1g1bko9-MuiButtonBase-root-MuiListItemButton-root"
    role="button"
    tabindex="0"
  >
    <div
      class="MuiListItemText-root MuiListItemText-multiline search-revision-item-text css-konndc-MuiListItemText-root"
    >
      <span
        class="MuiTypography-root MuiTypography-body1 MuiTypography-noWrap MuiListItemText-primary css-hrbm3k-MuiTypography-root"
      >
        <span
          class="MuiTypography-root MuiTypography-body2 revision-hash css-1lnwygr-MuiTypography-root"
        >
          <a
            class="MuiTypography-root MuiTypography-inherit MuiLink-root MuiLink-underlineAlways css-1qhous5-MuiTypography-root-MuiLink-root"
            href="https://treeherder.mozilla.org/jobs?repo=try&revision=coconut"
            target="_blank"
            title="open treeherder view for coconut"
          >
            coconut
          </a>
        </span>
        <div
          class="info-caption"
        >
          <div
            class="info-caption-item item-author"
          >
             
            <svg
              aria-hidden="true"
              class="MuiSvgIcon-root MuiSvgIcon-fontSizeSmall mail-icon css-1mecf1h-MuiSvgIcon-root"
              data-testid="MailOutlineOutlinedIcon"
              focusable="false"
              viewBox="0 0 24 24"
            >
              <path
                d="M20 4H4c-1.1 0-1.99.9-1.99 2L2 18c0 1.1.9 2 2 2h16c1.1 0 2-.9 2-2V6c0-1.1-.9-2-2-2zm0 14H4V8l8 5 8-5v10zm-8-7L4 6h16l-8 5z"
              />
            </svg>
             
            johncleese@python.com
          </div>
          <div
            class="info-caption-item item-time"
          >
            <svg
              aria-hidden="true"
              class="MuiSvgIcon-root MuiSvgIcon-fontSizeSmall time-icon css-1mecf1h-MuiSvgIcon-root"
              data-testid="AccessTimeOutlinedIcon"
              focusable="false"
              viewBox="0 0 24 24"
            >
              <path
                d="M11.99 2C6.47 2 2 6.48 2 12s4.47 10 9.99 10C17.52 22 22 17.52 22 12S17.52 2 11.99 2zM12 20c-4.42 0-8-3.58-8-8s3.58-8 8-8 8 3.58 8 8-3.58 8-8 8zm.5-13H11v6l5.25 3.15.75-1.23-4.5-2.67z"
              />
            </svg>
            03/29/51 00:00
          </div>
        </div>
      </span>
      <p
        class="MuiTypography-root MuiTypography-body2 MuiTypography-noWrap MuiListItemText-secondary css-163o6oo-MuiTypography-root"
      >
        you've got no arms left! 
      </p>
    </div>
    <button
      class="MuiButtonBase-root MuiButton-root MuiButton-text MuiButton-textPrimary MuiButton-sizeMedium MuiButton-textSizeMedium MuiButton-root MuiButton-text MuiButton-textPrimary MuiButton-sizeMedium MuiButton-textSizeMedium icon-close-show revision-action close-button css-13h131f-MuiButtonBase-root-MuiButton-root"
      name="close-button"
      tabindex="0"
      title="remove revision"
      type="button"
    >
      <svg
        aria-hidden="true"
        class="MuiSvgIcon-root MuiSvgIcon-fontSizeSmall css-1mecf1h-MuiSvgIcon-root"
        data-testid="close-icon"
        focusable="false"
        viewBox="0 0 24 24"
      >
        <path
          d="M19 6.41 17.59 5 12 10.59 6.41 5 5 6.41 10.59 12 5 17.59 6.41 19 12 13.41 17.59 19 19 17.59 13.41 12 19 6.41z"
        />
      </svg>
      <span
        class="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root"
      />
    </button>
    <span
      class="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root"
    />
  </div>
</li>
`;
